src/target/nrf51.c

```diff
@@ -103,87 +105,40 @@ static void nrf51_add_flash(target *t,
-	case 0x0072: /* nRF51822 (rev 3) QFAA H0 */
-	case 0x0073: /* nRF51422 (rev 3) QFAA F0 */
-	case 0x0079: /* nRF51822 (rev 3) CEAA E0 */
-	case 0x007A: /* nRF51422 (rev 3) CEAA C0 */
-	case 0x008F: /* nRF51822 (rev 3) QFAA H1 See https://devzone.nordicsemi.com/question/97769/can-someone-conform-the-config-id-code-for-the-nrf51822qfaah1/ */
-	case 0x00D1: /* nRF51822 (rev 3) QFAA H2 */
-	case 0x0114: /* nRF51802 (rev ?) QFAA A1 */
-		t->driver = "Nordic nRF51";
-		target_add_ram(t, 0x20000000, 0x4000);
-		nrf51_add_flash(t, 0x00000000, 0x40000, NRF51_PAGE_SIZE);
-		nrf51_add_flash(t, NRF51_UICR, 0x100, 0x100);
-		target_add_commands(t, nrf51_cmd_list, "nRF51");
-		return true;
-	case 0x0026: /* nRF51822 (rev 1) QFAB AA */
-	case 0x0027: /* nRF51822 (rev 1) QFAB A0 */
-	case 0x004C: /* nRF51822 (rev 2) QFAB B0 */
-	case 0x0061: /* nRF51422 (rev 2) QFAB A00 */
-	case 0x007B: /* nRF51822 (rev 3) QFAB C0 */
-	case 0x007C: /* nRF51422 (rev 3) QFAB B0 */
-	case 0x007D: /* nRF51822 (rev 3) CDAB A0 */
-	case 0x007E: /* nRF51422 (rev 3) CDAB A0 */
-		t->driver = "Nordic nRF51";
-		target_add_ram(t, 0x20000000, 0x4000);
-		nrf51_add_flash(t, 0x00000000, 0x20000, NRF51_PAGE_SIZE);
-		nrf51_add_flash(t, NRF51_UICR, 0x100, 0x100);
-		target_add_commands(t, nrf51_cmd_list, "nRF51");
+	uint32_t page_size = target_mem_read32(t, NRF51_FICR_CODEPAGESIZE);
+	uint32_t code_size = target_mem_read32(t, NRF51_FICR_CODESIZE);
+	/* Check that page_size and code_size makes sense */
+	if ((page_size == 0xffffffff) || (code_size == 0xffffffff) ||
+		(page_size ==  0) || (code_size ==  0) ||
+		(page_size > 0x10000) || (code_size > 0x10000))
+		return false;
+	/* Check that device identifier makes sense */
+	uint32_t uid0 = target_mem_read32(t, NRF51_FICR_DEVICEID_LOW);
+	uint32_t uid1 = target_mem_read32(t, NRF51_FICR_DEVICEID_HIGH);
+	if ((uid0 == 0xffffffff) || (uid1 == 0xffffffff) ||
+		(uid0 ==  0) || (uid1 ==  0))
+		return false;
+	/* Test for NRF52 device*/
+	uint32_t info_part = target_mem_read32(t, NRF52_PART_INFO);
+	if ((info_part != 0xffffffff) && (info_part != 0) &&
+		((info_part & 0x00ff000) == 0x52000)) {
+		uint32_t ram_size = target_mem_read32(t, NRF52_INFO_RAM);
+		t->idcode = info_part;
+		t->driver = "Nordic nRF52";
+		target_add_ram(t, 0x20000000, ram_size * 1024);
+		nrf51_add_flash(t, 0, page_size * code_size, page_size);
+		target_add_commands(t, nrf51_cmd_list, "nRF52");
 		return true;
-	case 0x0071: /* nRF51422 (rev 3) QFAC AB */
-	case 0x0083: /* nRF51822 (rev 3) QFAC A0 */
-	case 0x0084: /* nRF51422 (rev 3) QFAC A1 */
-	case 0x0085: /* nRF51422 (rev 3) QFAC A0 */
-	case 0x0086: /* nRF51422 (rev 3) QFAC A1 */
-	case 0x0087: /* nRF51822 (rev 3) CFAC A0 */
-	case 0x0088: /* nRF51422 (rev 3) CFAC A0 */
+	} else {
 		t->driver = "Nordic nRF51";
+		/* Use the biggest RAM size seen in NRF51 fammily.
+		 * IDCODE is kept as '0', as deciphering is hard and
+		 * there is later no usage.*/
 		target_add_ram(t, 0x20000000, 0x8000);
```

Comment:
This should use `NUMRAMBLOCK` and `SIZERAMBLOCKS` from the FICR on nRF51

Reply:
I am happy for it to use the largest possible ram size, rather than probing for the exact size.